### PR TITLE
Fix file upload inside file

### LINKED ISSUE #454: File upload inside a file
When we upload a file to some remotepath it creates parent directories if they already are not present. So when a file is uploaded say `/a.txt` blobber stores it as file type `f`, which is absolutely fine.

Now when we upload a file say `b.txt` with remotepath `/a.txt/b.txt` then blobber happily accepts it and when we do `zbox list --remotepath /a.txt` we can see `b.txt` in its list. And when we do `zbox list --remotepath /`, it shows `a.txt` with type `f`.

Blobber can provide this feature/issue, but it results in inconsistencies and lots of confusion for other tools and direct user as well.
We should not allow upload inside a file.

### PR DESCRIPTION
**Pull request by lpoli**
This PR fixes issue of file upload inside a file along with directory creation inside a file path

### CODE REVIEW

File: code/go/0chain.net/blobbercore/handler/file_command_add.go [hunk 1]
```diff
@@ -40,6 +40,8 @@ func (cmd *AddFileCommand) IsAuthorized(ctx context.Context, req *http.Request,
 
 	if exisitingFileRef != nil {
 		return common.NewError("duplicate_file", "File at path already exists")
+	} else if err := validateParentPathType(ctx, allocationObj.ID, fileChanger.Path); err != nil {
```
Comment (peterlimg) [inline code/go/0chain.net/blobbercore/handler/file_command_add.go]: We prefer do this in a new `if ` section instead of `if  .. else if..`.

File: code/go/0chain.net/blobbercore/reference/ref.go [hunk 2]
```diff
@@ -235,6 +235,59 @@ func GetReferenceFromLookupHash(ctx context.Context, allocationID, path_hash str
...
+		return nil, err
+	}
+	return ref, nil
+}
+
+// IsRefExist checks if ref with given path exists and returns error other than gorm.ErrRecordNotFound
+func IsRefExist(ctx context.Context, allocationID, path string) (bool, error) {
+	db := datastore.GetStore().GetTransaction(ctx)
+	ref := new(Ref)
+	err := db.Select("path").Where("allocation_id=? AND path=?", allocationID, path).First(ref).Error
```
Comment (peterlimg) [inline code/go/0chain.net/blobbercore/reference/ref.go]: Perhaps we can use db.Select("count (1) > 0") to avoid  unnecessary `ref` decoding.

File: code/go/0chain.net/core/common/utils.go [hunk 3]
```diff
@@ -22,3 +25,15 @@ func ToKey(key interface{}) string {
...
+// getParentPaths For path /a/b/c.txt, will return [/a,/a/b]
+func GetParentPaths(fPath string) []string {
+	splittedPaths := strings.Split(fPath, "/")
+	var paths []string
+
+	for i := 0; i < len(splittedPaths); i++ {
+		subPath := strings.Join(splittedPaths[0:i], "/")
+		paths = append(paths, subPath)
+	}
+	return paths[2:]
```
Comment (peterlimg) [inline code/go/0chain.net/core/common/utils.go]: This will panic if fPath is "". Please add unit tests to check the edge cases. 

Also, do you think we should consider the cases such as "/a//b", and //a//b, etc?

**Review by peterlimg — CHANGES_REQUESTED** [review]

File: code/go/0chain.net/core/common/utils.go (hunk 3 above)
Comment (lpoli) [reply code/go/0chain.net/core/common/utils.go]: I've added `filepath.Clean()` in this function and removed it from `validateParentType` function.

For above cases; I think its necessary to cleaning path before processing. It should be done in functions as early as relevant and we should check against if filepath is not absolute.

**Review by peterlimg — APPROVED** [review]
lgtm
